### PR TITLE
compute atomic self energies for a given data set prior to training 

### LINKED ISSUE #251: Include linear fitting for EnergyShifter
#243 

This has to be done on the training data

### PR DESCRIPTION
**Pull request by farhadrgh**
This pull request Fixes #251

### CODE REVIEW

File: torchani/utils.py [hunk 1]
```diff
@@ -142,8 +143,31 @@ class EnergyShifter(torch.nn.Module):
 
     def __init__(self, self_energies):
         super(EnergyShifter, self).__init__()
-        self_energies = torch.tensor(self_energies, dtype=torch.double)
+
+        if self_energies is not None:
+            self_energies = torch.tensor(self_energies, dtype=torch.double)
+
         self.register_buffer('self_energies', self_energies)
+        self.present_species = present_species
```
Comment (zasdfgbnm) [inline torchani/utils.py]: Why saving this function as a member?

File: torchani/utils.py [hunk 2]
```diff
@@ -142,8 +143,31 @@ class EnergyShifter(torch.nn.Module):
...
+        """Compute atomic self energies from dataset.
+
+        Least-squares solution to a linear equation is calculated to output
+        ``self_energies`` when ``self_energies = None`` is passed to
+        :class:`torchani.EnergyShifter`
+        """
+        species = atomic_properties['species']
+        energies = properties['energies']
+        present_species = self.present_species(species)
+        mask = []
```
Comment (zasdfgbnm) [inline torchani/utils.py]: Instead of having a `mask` and doing a for loop, could we just do the following:
```python
X = (species.unsqueeze(-1) == present_species).sum(dim=1).to(torch.double)
```

File: torchani/utils.py [hunk 3]
```diff
@@ -142,8 +143,31 @@ class EnergyShifter(torch.nn.Module):
...
         super(EnergyShifter, self).__init__()
-        self_energies = torch.tensor(self_energies, dtype=torch.double)
+
+        if self_energies is not None:
+            self_energies = torch.tensor(self_energies, dtype=torch.double)
+
         self.register_buffer('self_energies', self_energies)
+        self.present_species = present_species
+
+    def sae_from_dataset(self, atomic_properties, properties):
```
Comment (zasdfgbnm) [inline torchani/utils.py]: Could you make this a `@staticmethod`? No member function is used in this function.

Comment (farhadrgh) [reply torchani/utils.py]: Fixed in dc85d3c

File: torchani/utils.py (hunk 2 above)
Comment (farhadrgh) [reply torchani/utils.py]: Fixed in dc85d3c

File: torchani/utils.py (hunk 1 above)
Comment (farhadrgh) [reply torchani/utils.py]: Fixed in dc85d3c
